### DISCUSSION TITLE
Metadata search (again, sorry!)

### DISCUSSION THREAD

**Discussion by WS64** [koreader/koreader#7103] [Q&A] [answered]
I guess this goes mainly to @pazos!

Hi, still trying to get the metadata search to run.
(for those who have not read #7081, "find a book my calibre metadata" does not work when the metadata are too big)

I thought I try to solve that issue for me from a different angle by creating ./cache/calibre-books.lua by myself instead of letting Koreader doing it.
I have a solution (not 100% perfect yet, I do it with AWK and have a problem with the UTF-16 characters, but good enough to test it already). So I let Koreader search for a manipulated small metdata.calibre file and did not touch it anymore.
I checked the resulting calibre-books.lua and created my own with the 50000 books with the above mentioned AWK program.
Of course I made sure this calibre-books.lua got a newer timestamp than the metadata file to make sure Koreader will not re-create it.

So far so good, but crash.log tells me...

`01/01/21-21:40:47 WARN  invalid cache: ./cache/calibre-books.lua:488526: main function has more than 65536 constants`

The created calibre-books.lua has about 50000 entries (488526 lines), each of the entries looks like this:
    [8] = {
        ["tags"] = {
            [1] = "Science Fiction",
            [2] = "read",
        },
        ["lpath"] = "A/Aaron Allston/Conviction.epub",
        ["rootpath"] = "/mnt/ws/.hidden",
        ["authors"] = {
            [1] = "Aaron Allston",
        },
        ["title"] = "Conviction",
    },
Is there a way to get around this?

And additional question, if this is somehow solveable, the whole code to create this file actually is already there in \plugins\calibre.koplugin (and some dependencies). Would it be possible to extract it from Koreader to create a stand-alone lua progam that does nothing but create calibre-books.lua from an existing ,metadata.calibre on my Windows PC? I tried to get it working but I fail already at the very first steps to run metadata.lua, no idea how to include rapidjson and the other requirements! (The only time I did anything in lua was 6 years ago when I did some parts, including the original metadata search function)

**Comment by pazos** [comment]
See http://lua-users.org/lists/lua-l/2010-03/msg00238.html

Each book is a table with a few tables inside. I didn't know these limits because I was miles away of them :/ but you can reach them with more than 20000 books, it seems.

So nothing can be done with the implementation as it is. It needs serious rework (ie: by spliting loadTable function, that trips the warning) or, much much better, to store the book metadata in a sqlite DB.

**Comment by pazos** [comment]
@WS64: could you attach the `calibre-books.lua` somewhere?. I'll see if I can use a serializer instead of `dofile` directly.

**Reply by WS64** [reply]
[test.zip](https://github.com/koreader/koreader/files/5759848/test.zip)
Here it is! (test.txt has to be renamed of course)
It still might contain errors, I was in the testing phase when I reached the limitation error.

**Comment by pazos** [comment]
The serializer approach works fine.

```lua
#!/usr/bin/luajit

local blob = require("blob")

-- sample book table
local function addBook(index)
   local n = index or 0
   return {
       ["tags"] = {
           [1] = "Book kind",
           [2] = "Book kind 2",
       },
       ["lpath"] = "relpath/to/book" .. n,
       ["rootpath"] = "/something/something",
       ["authors"] = {
           [1] = "Author 1",
           [2] = "Author 2",
       },
       ["title"] = "book" .. n,
   }
end

-- create a table with enough nested tables to trigger
-- http://lua-users.org/lists/lua-l/2010-03/msg00238.html
-- that's 65536 + 1
local function gen()
    local t = {}
    for i = 1, 100000 do
        table.insert(t, i, addBook(i))
    end
    return t
end

-- dump the table to file
blob.saveBlob("books.cache", gen())

-- load the table from file
local t = blob.loadBlob("books.cache")
```

So, if you manage to generate a table with books, no matter how large it is, we can save to file and load from file just fine.

But **you need** to do the table creation in lua and dump from there. It is not possible to load a previously generated table from file if it has more than 65536 nested tables.

**Reply by Frenzie** [reply]
Hey, I know that number (or something much like it) as a limit from Excel! :-D 

https://www.mrexcel.com/tech-tv/overcome-excels-65536-limit/

**Reply by WS64** [reply]
@pazos, where does that code go to?

@Frenzie, it's called 64k! ;)

**Reply by pazos** [reply]
The code is just boilerplate that creates a table with 100000 books, serializes it and dumps it to a file. It is able to retrieve the same table from file without problems.

So, it goes nowhere, but the code

```lua
local binser = require("binser")

local Blob = {}

function Blob.saveBlob(file, t)
    local f, err = io.open(file, "w")
    if not f then return nil, err end
    f:write(binser.serialize(t))
    f:close()
    return true
end

function Blob.loadBlob(file)
    local f, err = io.open(file, "r")
    if not f then return nil, err end
    local t = binser.deserializeN(f:read "*a")
    f:close()
    return t
end

return Blob
```

can be used instead of [`loadTable`](https://github.com/koreader/koreader/blob/master/plugins/calibre.koplugin/search.lua#L28-L36) and [`util.dumpTable`](https://github.com/koreader/koreader/blob/master/plugins/calibre.koplugin/search.lua#L615)

to be able to have persistence with huge tables (a million books would mean a binary of 60MB or so)

It still doesn't solve the OOM issue importing from json. But once imported (by using a SAX enabled json lib or by manually parsing the text, like you were doing) it can be saved/loaded no matter how big the table is.

**Comment by pazos** [comment]
@WS64: wrote a raw parser based on your previous code:

```lua
local util = require("util")

local function isOk(file)
    local f, err = io.open(file, "rb")
    if f then f:close() end
    return f and f or nil, err
end

-- removes leading and closing characters and converts hex-unicodes
local function replaceHexChars(s, n, j)
    local l = string.len(s)
    if string.sub(s, l, l) == "\"" then
        s = string.sub(s, n, string.len(s)-1)
    else
        s = string.sub(s, n, string.len(s)-j)
    end
    s = string.gsub(s, "\\u([a-f0-9][a-f0-9][a-f0-9][a-f0-9])", function(w)
        return util.unicodeCodepointToUtf8(tonumber(w, 16)) 
    end)
    return s
end

local function add(line, t)
    line = replaceHexChars(line, 8, 3)
    if type(t) ~= "table" then
        return {}
    end
    local tab = t or {}
    if type(line) == "string" then
        table.insert(tab, #tab + 1, line)
    end
    return tab
end

local function parse(file)
    local ok, err = isOk(file)
    if not ok then
        return nil, err
    end
    local books = {}
    local book_entry = {}
    local is_author = false
    local is_tag = false

    for line in io.lines(file) do
        if type(line) == "string" then
            if line == "  }, " or line == "  }" then
                if type(book_entry) == "table" then
                    table.insert(books, #books + 1, book_entry)
                end
                book_entry = {}
            elseif line == "    \"authors\": [" then
                is_author = true
            elseif line == "    \"tags\": [" then
                is_tag = true
            elseif string.sub(line, 1, 11) == "    \"title\"" then
                book_entry.title = replaceHexChars(line, 15, 3)
            elseif string.sub(line, 1, 11) == "    \"lpath\"" then
                book_entry.lpath = replaceHexChars(line, 15, 3)
            elseif string.sub(line, 1, 10) == "    \"uuid\"" then
                book_entry.uuid = replaceHexChars(line, 14, 3)
            elseif string.sub(line, 1, 12) == "    \"series\"" and line ~= "    \"series\": null, " then -- SERIES
                book_entry.series = replaceHexChars(line, 16, 3)
            elseif string.sub(line, 1, 18) == "    \"series_index\"" and line ~= "    \"series_index\": null, " then -- SERIES_INDEX
                book_entry.series_index = replaceHexChars(line, 21, 2)
            elseif line == "    ], " or line == "    ]" then
                is_author = false
                is_tag = false
            end

            if is_author then
                book_entry.authors = add(line, book_entry.authors)
            elseif is_tag then
                book_entry.tags = add(line, book_entry.tags)
            end
        end
    end
    return books
end

local RawParser = {}

function RawParser.fromFile(file)
    if type(file) ~= "string" then return {} end
    print("RawParser: loading table from file " .. file)
    local t, err = parse(file)
    if not t then
        print("RawParser: " .. err)
        return {}
    end
    return t
end

return RawParser
```

Applied to your `.metadata.calibre` it returns a table with 50252 books. Each book looks like:

```lua
    [50251] = {
        ["lpath"] = "D/Diverse/The Science Fiction Hall of Fame Volume 1.epub",
        ["title"] = "The Science Fiction Hall of Fame Volume 1",
        ["series"] = "Science Fiction Hall of Fame",
        ["tags"] = {
            [1] = "Science Fiction",
            [2] = "Shorts",
        },
        ["series_index"] = "1.0",
        ["uuid"] = "b4f4192f-bef1-4ea4-97b3-ca9c451e4021",
        ["authors"] = {
            [1] = "Diverse",
        },
    },
    [50252] = {
        ["title"] = "Microcosmic Tales",
        ["lpath"] = "D/Diverse/Microcosmic Tales.epub",
        ["uuid"] = "a16fda6f-224c-45dc-9129-0a944b0cb809",
        ["authors"] = {
            [1] = "Diverse",
        },
    },
```

On my -no so fast computer - takes a couple of seconds to parse the whole file and generate the table. On your Nia one minute or so is reasonable, but I think it will end sooner.

**Reply by WS64** [reply]
Thanks, @Pazos.
How can I run this on my Kobo?

**Reply by pazos** [reply]
I need a couple of free hours to put all together in the calibre plugin. Probably I'll be able to do it next weekend (no holidays here :disappointed:)

In the mean time I guess you can hack around your way with the code pasted here to know how well it behaves on your device.

My tests with your metadata show:

```
generating cache for metadata-huge.json (179352kB)
done in 2463.732958 milliseconds
dumping cache in metadata-huge.cache
done in 327.986956 milliseconds
loading cache from metadata-huge.cache
done in 148.533106 milliseconds
```

The resulting `metadata-huge.cache` is a binary file of 6.4MB. Once generated it can be loaded in about 150ms. I guess the Nia will be quite limited by IO, so timmings of 10x - 20x are feasible. Once loaded iterating the table for lookups would be quite fast (usually 0.1x the time of loading the data)

**Reply by pazos** [reply]
Ok, 2 hours was a bit too much. I ended up pretty quickly.

Sadly I cannot test more without these 50000 books, as the code that loads the metadata checks that each book is there, which is not true in my case.

Anyways, the result is:

```
01/05/21-18:47:46 INFO  calibre info loaded from disk in 30907.371998 milliseconds: 25126 books. 25126 pruned
01/05/21-18:47:46 INFO  calibre info loaded from disk in 0.099897 milliseconds: 1 books. 0 pruned
```

Most of the time spent was prunning books from the table, so it should be faster with books actually there. @WS64 Please test (replace the calibre.koplugin with the contents of 
[calibre.koplugin.zip](https://github.com/koreader/koreader/files/5771868/calibre.koplugin.zip))

Do a backup first

**Reply by pazos** [reply]
BTW, it seems that most of your books are actually duplicated, as the `pruner` finds 25126 books and prunes all of them.

**Reply by WS64** [reply]
Thanks, but:


01/05/21-19:19:37 INFO  FM loaded plugin terminal at plugins/terminal.koplugin
01/05/21-19:19:37 INFO  FM loaded plugin texteditor at plugins/texteditor.koplugin
01/05/21-19:19:37 INFO  FM loaded plugin timesync at plugins/timesync.koplugin
01/05/21-19:19:37 INFO  FM loaded plugin wallabag at plugins/wallabag.koplugin
01/05/21-19:19:50 INFO  event times are real clock time: no adjustment needed
01/05/21-19:20:12 WARN  no calibre libraries cannot open ./cache/calibre-libraries.lua: No such file or directory
./luajit: plugins/calibre.koplugin/search.lua:572: attempt to call field 'loadFile' (a nil value)
stack traceback:
	plugins/calibre.koplugin/search.lua:572: in function 'getMetadata'
	plugins/calibre.koplugin/search.lua:502: in function 'ok_callback'
	frontend/ui/widget/confirmbox.lua:100: in function 'callback'
	frontend/ui/widget/button.lua:239: in function 'action'
	frontend/ui/uimanager.lua:882: in function '_checkTasks'
	frontend/ui/uimanager.lua:1219: in function 'handleInput'
	frontend/ui/uimanager.lua:1312: in function 'run'
	./reader.lua:300: in main chunk
	[C]: at 0x00014131
!!!!
Uh oh, something went awry... (Crash n°1: 01/05/21 @ 19:22:15)
Running FW 4.22.15268 on Linux 4.1.15-00437-g3064755b04b9 (#11 SMP PREEMPT Tue Jul 7 08:35:57 CST 2020)
Attempting to restart KOReader . . .

**Reply by pazos** [reply]
That's weird. That line should be

`local ok, err = cache.loadFile(self.user_book_cache)`

**Reply by pazos** [reply]
try with this one 
[calibre.koplugin.zip](https://github.com/koreader/koreader/files/5771949/calibre.koplugin.zip)

**Reply by WS64** [reply]
That one crashed immediately (the first one took 2 minutees or so):

01/05/21-19:57:36 INFO  FM loaded plugin texteditor at plugins/texteditor.koplugin
01/05/21-19:57:36 INFO  FM loaded plugin timesync at plugins/timesync.koplugin
01/05/21-19:57:36 INFO  FM loaded plugin wallabag at plugins/wallabag.koplugin
01/05/21-19:57:39 INFO  event times are real clock time: no adjustment needed
./luajit: plugins/calibre.koplugin/search.lua:572: attempt to call field 'loadFile' (a nil value)
stack traceback:
	plugins/calibre.koplugin/search.lua:572: in function 'getMetadata'
	plugins/calibre.koplugin/search.lua:318: in function 'find'
	plugins/calibre.koplugin/search.lua:241: in function 'close'
	plugins/calibre.koplugin/search.lua:224: in function 'callback'
	frontend/ui/widget/button.lua:239: in function 'action'
	frontend/ui/uimanager.lua:882: in function '_checkTasks'
	frontend/ui/uimanager.lua:1219: in function 'handleInput'
	frontend/ui/uimanager.lua:1312: in function 'run'
	./reader.lua:300: in main chunk
	[C]: at 0x00014131
!!!!
Uh oh, something went awry... (Crash n°1: 01/05/21 @ 19:57:50)
Running FW 4.22.15268 on Linux 4.1.15-00437-g3064755b04b9 (#11 SMP PREEMPT Tue Jul 7 08:35:57 CST 2020)
Attempting to restart KOReader . . .
!!!!

**Reply by WS64** [reply]
@pazos, any news here?

**Reply by pazos** [reply]
Yeah!. But I factorized a bit the update because your requirements need various improvements on the cache stuff. See: https://github.com/koreader/koreader/pull/7120

Also, I worked a little bit on intercompatibility between the raw parser (based on your previous contributions) and the json parser, so both should be able to work transparently and push & pop books between them.

I'll push the changes needed on the parser this afternoon

**Comment by pazos** [comment, marked as answer]
@WS64: It is live on https://github.com/koreader/koreader/pull/7159. Not ready for public comsuption but ok for you to test if you wish.

**Reply by WS64** [reply]
Might be a stupid question, but how do I download #7159 from the web interface?

**Reply by Frenzie** [reply]
Since it's just one file you can use changes → view file and then there's also something like "raw".

**Reply by WS64** [reply]
I have replaced metadata.lua with the raw data from #7159.
But, I cannot find the menu option anymore at all to search in the calbre database. Where was it?
I cannot test :(

**Reply by pazos** [reply]
you need to copy `parser.lua` and `search.lua` too. And apply all that on a current nightly (something in last week should work)

**Reply by WS64** [reply]
Thanks, got it working!
First search took about 2 minutes, fromthen on it is fast (as expected).
One last question, where do I set the font for the search results? Hard to read, much too small...

**Reply by pazos** [reply]
> First search took about 2 minutes

Yup. That means your Nia storage at least 2x slower than your old Aura. That makes sense as the old had an internal emmc and the Nia a crappy sdcard.

> One last question, where do I set the font for the search results? Hard to read, much too small

No idea, i didn't change that and you should expect the same look & feel as always (and in consonance with the rest of the UI)
